### ISSUE TITLE
Optimize data augmentation for 3D kernel

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/model_seg_ms_mp2rage#57]
### Useful syntax

To look at predictions with GT. Run under the `data_seg_mp2rage_20221220_124553/data_processed_lesionseg` folder:
~~~
for i in 017 025 058 079 082 098 103 111 114 115; do fsleyes -S sub-P${i}/anat/sub-P${i}_UNIT1 derivatives/labels/sub-P${i}/anat/sub-P${i}_UNIT1_lesion-manualHaris -cm yellow ../../model_seg_lesion_mp2rage_xxx/pred_masks/sub-P${i}_UNIT1_pred -cm red -a 50; done
~~~

TODO:
- [x] Use less conservative cropping during preprocessing, to avoid the 'black edge', which could limit the performance of training (ie: model would focus on edge detection).
- [ ] Use softseg
- [ ] Looks like only one lesion is segmented. Maybe check post-processing (keep only one object).

![image](https://user-images.githubusercontent.com/2482071/208268199-d6ff9037-e4a9-4abc-98d5-c1b45f7c848a.png)
<details><summary>config file</summary>

```console
{
    "command": "train",
    "gpu_ids": [4],
    "path_output": "/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/model_seg_lesion_mp2rage_",
    "model_name": "model_seg_lesion_mp2rage",
    "debugging": true,
    "object_detection_params": {
        "object_detection_path": null,
        "safety_factor": [1.0, 1.0, 1.0]
    },
    "wandb": {
        "wandb_api_key": "9095e2bc9e4ab445d478c9c8a81759ae908be8c6",
        "project_name": "basel-mp2rage-lesion",
        "group_name": "3D",
        "run_name": "run-1",
        "log_grads_every": 100
    },
    "loader_parameters": {
        "path_data": ["/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/data_seg_mp2rage_20221217_170634/data_processed_lesionseg"],
        "subject_selection": {"n": [], "metadata": [], "value": []},
        "target_suffix": ["_lesion-manualHaris"],
        "extensions": [".nii.gz"],
        "roi_params": {
            "suffix": null,
            "slice_filter_roi": null
        },
        "contrast_params": {
            "training_validation": ["UNIT1"],
            "testing": ["UNIT1"],
            "balance": {}
        },
        "slice_filter_params": {
            "filter_empty_mask": true,
            "filter_empty_input": true
        },
        "slice_axis": "axial",
        "multichannel": false,
        "soft_gt": false,
        "bids_validate": true
    },
    "split_dataset": {
        "fname_split": null,
        "random_seed": 42,
        "split_method" : "participant_id",
        "data_testing": {"data_type": null, "data_value":[]},
        "balance": null,
        "train_fraction": 0.6,
        "test_fraction": 0.2
    },
    "training_parameters": {
        "batch_size": 16,
        "loss": {
            "name": "DiceLoss"
        },
        "training_time": {
            "num_epochs": 50,
            "early_stopping_patience": 50,
            "early_stopping_epsilon": 0.001
        },
        "scheduler": {
            "initial_lr": 0.001,
            "lr_scheduler": {
                "name": "CosineAnnealingLR",
                "base_lr": 1e-5,
                "max_lr": 1e-3
            }
        },
        "balance_samples": {
            "applied": false,
            "type": "gt"
        },
        "mixup_alpha": null,
        "transfer_learning": {
            "retrain_model": null,
            "retrain_fraction": 1.0,
            "reset": true
        }
    },
    "default_model": {
        "name": "Unet",
        "dropout_rate": 0.3,
        "bn_momentum": 0.1,
        "final_activation": "sigmoid",
        "depth": 3
    },
    "FiLMedUnet": {
        "applied": false,
        "metadata": "contrasts",
        "film_layers": [0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    "Modified3DUNet": {
        "applied": true,
        "length_3D": [32, 32, 64],
        "stride_3D": [32, 32, 64],
        "attention": false,
        "n_filters": 8
    },
    "uncertainty": {
        "epistemic": false,
        "aleatoric": false,
        "n_it": 0
    },
    "postprocessing": {
        "remove_noise": {"thr": -1},
        "keep_largest": {},
        "binarize_prediction": {"thr": 0.5},
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "fill_holes": {},
        "remove_small": {"unit": "vox", "thr": 3}
    },
    "evaluation_parameters": {
        "target_size": {"unit": "vox", "thr": [20, 100]},
        "overlap": {"unit": "vox", "thr": 3}
    },
    "transformation": {
        "Resample":
        {
            "hspace": 0.75,
            "wspace": 0.75,
            "dspace": 0.75
        },
        "CenterCrop": {
            "size": [64, 64, 128]},

        "RandomAffine": {
            "degrees": 5,
            "scale": [0.1, 0.1],
            "translate": [0.1, 0.1],
            "applied_to": ["im", "gt"],
            "dataset_type": ["training"]
        },
        "ElasticTransform": {
            "alpha_range": [28.0, 30.0],
            "sigma_range":  [3.5, 4.5],
            "p": 0.1,
            "applied_to": ["im", "gt"],
            "dataset_type": ["training"]
	},		    
	"NormalizeInstance": {"applied_to": ["im"]}
    }
}
```

</details>

**Comment by jcohenadad** [comment]
![image](https://user-images.githubusercontent.com/2482071/208268410-ac04b141-4b0e-4109-8de0-e72a4208938a.png)

<details><summary>Terminal output</summary>

```console
{
    "command": "train",
    "gpu_ids": [4],
    "path_output": "/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/model_seg_lesion_mp2rage_",
    "model_name": "model_seg_lesion_mp2rage",
    "debugging": true,
    "object_detection_params": {
        "object_detection_path": null,
        "safety_factor": [1.0, 1.0, 1.0]
    },
    "wandb": {
        "wandb_api_key": "9095e2bc9e4ab445d478c9c8a81759ae908be8c6",
        "project_name": "basel-mp2rage-lesion",
        "group_name": "3D",
        "run_name": "run-1",
        "log_grads_every": 100
    },
    "loader_parameters": {
        "path_data": ["/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/data_seg_mp2rage_20221217_170634/data_processed_lesionseg"],
        "subject_selection": {"n": [], "metadata": [], "value": []},
        "target_suffix": ["_lesion-manualHaris"],
        "extensions": [".nii.gz"],
        "roi_params": {
            "suffix": null,
            "slice_filter_roi": null
        },
        "contrast_params": {
            "training_validation": ["UNIT1"],
            "testing": ["UNIT1"],
            "balance": {}
        },
        "slice_filter_params": {
            "filter_empty_mask": true,
            "filter_empty_input": true
        },
        "slice_axis": "axial",
        "multichannel": false,
        "soft_gt": false,
        "bids_validate": true
    },
    "split_dataset": {
        "fname_split": null,
        "random_seed": 42,
        "split_method" : "participant_id",
        "data_testing": {"data_type": null, "data_value":[]},
        "balance": null,
        "train_fraction": 0.6,
        "test_fraction": 0.2
    },
    "training_parameters": {
        "batch_size": 16,
        "loss": {
            "name": "DiceLoss"
        },
        "training_time": {
            "num_epochs": 50,
            "early_stopping_patience": 50,
            "early_stopping_epsilon": 0.001
        },
        "scheduler": {
            "initial_lr": 0.001,
            "lr_scheduler": {
                "name": "CosineAnnealingLR",
                "base_lr": 1e-5,
                "max_lr": 1e-3
            }
        },
        "balance_samples": {
            "applied": false,
            "type": "gt"
        },
        "mixup_alpha": null,
        "transfer_learning": {
            "retrain_model": null,
            "retrain_fraction": 1.0,
            "reset": true
        }
    },
    "default_model": {
        "name": "Unet",
        "dropout_rate": 0.3,
        "bn_momentum": 0.1,
        "final_activation": "sigmoid",
        "depth": 3
    },
    "FiLMedUnet": {
        "applied": false,
        "metadata": "contrasts",
        "film_layers": [0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    "Modified3DUNet": {
        "applied": true,
        "length_3D": [32, 32, 64],
        "stride_3D": [32, 32, 64],
        "attention": false,
        "n_filters": 8
    },
    "uncertainty": {
        "epistemic": false,
        "aleatoric": false,
        "n_it": 0
    },
    "postprocessing": {
        "remove_noise": {"thr": -1},
        "keep_largest": {},
        "binarize_prediction": {"thr": 0.5},
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "fill_holes": {},
        "remove_small": {"unit": "vox", "thr": 3}
    },
    "evaluation_parameters": {
        "target_size": {"unit": "vox", "thr": [20, 100]},
        "overlap": {"unit": "vox", "thr": 3}
    },
    "transformation": {
        "Resample":
        {
            "hspace": 0.75,
            "wspace": 0.75,
            "dspace": 0.75
        },
        "CenterCrop": {
            "size": [128, 128, 128]},
	"NormalizeInstance": {"applied_to": ["im"]}
    }
}

```

</details>

**Comment by jcohenadad** [comment]
![image](https://user-images.githubusercontent.com/2482071/208268503-4f763ce9-9846-49fe-94bc-24adde2909e1.png)

Change:
~~~
        {
            "hspace": 1,
            "wspace": 1,
            "dspace": 1
        },
        "CenterCrop": {
            "size": [64, 64, 64]},
~~~

**Comment by jcohenadad** [comment]
Investigating the orientation and kernel size. With the following config:

```json
        "slice_axis": "axial",
...
    "Modified3DUNet": {
        "applied": true,
        "length_3D": [
            32,
            32,
            16
        ],
```

Here is a patch:

```python
imgplot = plt.imshow(img[4, 0, :, :, 8])
```

<img width="752" alt="Screen Shot 2023-01-04 at 2 33 10 PM" src="https://user-images.githubusercontent.com/2482071/210634936-f0375206-4f52-4c24-990a-f660041b0888.png">

So, it seems like the orientation is consistent, ie: the superior-inferior direction is along the 3rd dim (the "8").

**Comment by jcohenadad** [comment]
Now, let's look whether a modified CenterCrop enables to reach the entire SI axis:

```JSON
        "CenterCrop": {"size": [32, 32, 128]},
```

Provides the proper coverage. Below are one image corresponding to the center of each patch (16 vox in SI direction), using stride of 14 along the SI direction (hence, there are 9 batches to cover 128):
<img width="708" alt="Screen Shot 2023-01-04 at 8 05 12 PM" src="https://user-images.githubusercontent.com/2482071/210679323-e09b5159-56d4-4c2d-8204-db55176a1b00.png">
<img width="708" alt="Screen Shot 2023-01-04 at 8 06 03 PM" src="https://user-images.githubusercontent.com/2482071/210679325-ec2efc81-f79f-4d3c-8fac-b461ba34c2c1.png">
<img width="752" alt="Screen Shot 2023-01-04 at 8 06 17 PM" src="https://user-images.githubusercontent.com/2482071/210679326-853e83bf-433b-4496-af53-afda29711b6f.png">
<img width="708" alt="Screen Shot 2023-01-04 at 8 06 31 PM" src="https://user-images.githubusercontent.com/2482071/210679328-a6e0dfd9-01f1-44db-bb09-fa604d2bf53c.png">
<img width="708" alt="Screen Shot 2023-01-04 at 8 06 41 PM" src="https://user-images.githubusercontent.com/2482071/210679332-3aebcea1-4518-4776-999d-2955044d64f5.png">
<img width="708" alt="Screen Shot 2023-01-04 at 8 06 49 PM" src="https://user-images.githubusercontent.com/2482071/210679334-a882b892-6867-47d3-aa58-4412fe1afa70.png">
<img width="708" alt="Screen Shot 2023-01-04 at 8 06 58 PM" src="https://user-images.githubusercontent.com/2482071/210679335-4473c851-55be-4b1d-88c5-ffa121384d3f.png">
<img width="708" alt="Screen Shot 2023-01-04 at 8 07 17 PM" src="https://user-images.githubusercontent.com/2482071/210679337-ea1563e5-7657-4e3d-a629-b99bdaefa203.png">

One thing to note, though, is that the centercrop along the AP direction might be too aggressive.

**Comment by jcohenadad** [comment]
This is how a CenterCrop below look like:
```json
        "CenterCrop": {"size": [64, 32, 128]},
```

<img width="752" alt="Screen Shot 2023-01-04 at 8 35 59 PM" src="https://user-images.githubusercontent.com/2482071/210681898-30ec5734-4ef8-44ce-8e98-4c60458497d7.png">


So we need to go with 32x64x128.

**Comment by jcohenadad** [comment]
OK well, 64x128x96 is not appropriate:
![image](https://user-images.githubusercontent.com/2482071/210685051-a5365872-0a53-412a-bba7-a6c4f49b3893.png)

**Comment by jcohenadad** [comment]
that seems more appropriate, but I don't like the cropping in the AP direction with:
```json
        "length_3D": [32, 32, 64],
        "stride_3D": [1, 2, 2],
        "CenterCrop": {"size": [32, 64, 128]},
```

![image](https://user-images.githubusercontent.com/2482071/210685927-33448b9b-835a-41cd-ab20-96d092a0419e.png)

**Comment by jcohenadad** [comment]
Changing for:
```json
        "length_3D": [32, 64, 64],
        "stride_3D": [1, 1, 2],
        "CenterCrop": {"size": [32, 64, 128]},
```

Looks much better:
![image](https://user-images.githubusercontent.com/2482071/210687059-252c83e4-6e82-4c65-827c-53dbdad45bf7.png)
